plugins/synapticsmst/synapticsmst-device.c

```diff
@@ -504,221 +532,833 @@ synapticsmst_device_get_flash_checksum (SynapticsMSTDevice *device,
+	if (isChipPanamera == TRUE){
+		
+		g_debug("bypass check CusID %x %x",tmp, synapticsmst_device_get_board_id (device));	
 	}
-
+	else {
+		if (tmp != synapticsmst_device_get_board_id (device)) {
+			g_set_error_literal (error,	
+						G_IO_ERROR,	
+						G_IO_ERROR_INVALID_DATA,
+						"board ID mismatch");
+			return FALSE;
+		}
+	}
+	/* open device */
 	if (!synapticsmst_device_open (device, error)) {
 		g_prefix_error (error,
 				"can't open DP Aux node %s",
 				synapticsmst_device_get_aux_node (device));
 		return FALSE;
 	}
-
 	/* enable remote control */
 	if (!synapticsmst_device_enable_remote_control (device, error))
 		return FALSE;
 
-	/* erase SPI flash */
-	connection = synapticsmst_common_new (priv->fd, priv->layer, priv->rad);
-	if (synapticsmst_common_rc_set_command (connection,
-						UPDC_FLASH_ERASE,
-						2, 0, (guint8 *)&erase_code)) {
-		g_set_error_literal (error,
-				     G_IO_ERROR,
-				     G_IO_ERROR_INVALID_DATA,
-				     "can't erase flash");
-		return FALSE;
+	/* update firmware */
+	if (isChipPanamera == TRUE){
+
+		/* Need to detect flash mode and ESM first ? */
+		/* disable flash Quad mode and ESM/HDCP2.2*/
+		memset(dwData, 0, sizeof(dwData));			
+		rc = 0;
+		connection = synapticsmst_common_new (priv->fd, priv->layer, priv->rad);		
+		/* disable ESM first */
+		dwData[0] = 0x21;
+		rc = synapticsmst_common_rc_set_command (connection, 
+								UPDC_WRITE_TO_MEMORY,
+								4, (gint)REG_ESM_DISABLE, (guint8*)dwData);
+		g_usleep(200);/* waiting for ESM exit */
+		/* disable QUAD mode */ 
+		rc = synapticsmst_common_rc_get_command (connection, 
+								UPDC_READ_FROM_MEMORY,
+								((sizeof(dwData)/sizeof(dwData[0]))*4), 
+								(gint)REG_QUAD_DISABLE, (guint8*)dwData);		
+		dwData[0] = 0x00;
+		rc = synapticsmst_common_rc_set_command (connection, 
+								UPDC_WRITE_TO_MEMORY,
+								4, (gint)REG_QUAD_DISABLE, (guint8*)dwData);	
+		/* disable HDCP2.2 */
+		rc = synapticsmst_common_rc_get_command (connection, 
+								UPDC_READ_FROM_MEMORY,
+								4, (gint)REG_HDCP22_DISABLE, (guint8*)dwData);		
+		dwData[ 0 ] = dwData[0] & (~BIT(2));
+		rc = synapticsmst_common_rc_set_command (connection, 
+								UPDC_WRITE_TO_MEMORY,
+								4, (gint)REG_HDCP22_DISABLE, (guint8*)dwData);		
+		if (rc)	{							
+			g_set_error_literal (error, 
+						G_IO_ERROR, 
+						G_IO_ERROR_INVALID_DATA,
+						"can't disable Quad & HDCP");		
+			return FALSE;			
+		}
+		if (!synapticsmst_device_update_ESM(device, payload_data, error)){								
+			g_set_error_literal (error, 
+							G_IO_ERROR, 
+							G_IO_ERROR_INVALID_DATA,
+							"Update ESM fail");		
+			return FALSE;
+		}
+
+		if (!synapticsmst_device_update_Panamera_firmware(device, payload_len ,payload_data, error)){
+			g_set_error_literal (error, 
+							G_IO_ERROR, 
+							G_IO_ERROR_INVALID_DATA,
+							"Update firmware fail");		
+			return FALSE;
+		}
 	}
+	else
+	{ 
+		connection = synapticsmst_common_new (priv->fd, priv->layer, priv->rad);
```

Comment:
Shouldn't this block just split off to a separate function `synapticsmst_device_update_Leaf_firmware`?  to match style of panamera write?